File: poker_arena_server/src/game/PlayerManager.ts
import { Player, PlayerInfo } from './Player';
import { Logger } from '../utils/Logger';

/**
 * Player Manager (Server-side)
 * 管理游戏中的所有玩家（纯数据层，无UI）
 */
export class PlayerManager {
    private players: Map<string, Player> = new Map();
    private playerOrder: string[] = [];

    /**
     * 创建玩家
     */
    public createPlayers<T extends Player>(
        playerInfos: PlayerInfo[],
        PlayerClass: new (info: PlayerInfo) => T
    ): void {
        this.clear();

        for (const info of playerInfos) {
            const player = new PlayerClass(info);
            this.players.set(player.id, player);
            this.playerOrder.push(player.id);
        }

        Logger.info('PlayerManager', `Created ${this.players.size} players`);
    }

    /**
     * 获取玩家
     */
    public getPlayer(playerId: string): Player | undefined {
        return this.players.get(playerId);
    }

    /**
     * 获取所有玩家
     */
    public getAllPlayers(): Player[] {
        return Array.from(this.players.values());
    }

    /**
     * 获取玩家顺序
     */
    public getPlayerOrder(): string[] {
        return [...this.playerOrder];
    }

    /**
     * 获取玩家数量
     */
    public getPlayerCount(): number {
        return this.players.size;
    }

    /**
     * 获取玩家索引
     */
    public getPlayerIndex(playerId: string): number {
        return this.playerOrder.indexOf(playerId);
    }

    /**
     * 清空所有玩家
     */
    public clear(): void {
        this.players.clear();
        this.playerOrder = [];
    }

    /**
     * 更新玩家 ID（用于重连场景）
     * @param oldPlayerId 旧的玩家ID
     * @param newPlayerId 新的玩家ID
     * @returns 是否更新成功
     */
    public updatePlayerId(oldPlayerId: string, newPlayerId: string): boolean {
        const player = this.players.get(oldPlayerId);
        if (!player) {
            return false;
        }

        // 从旧的 key 删除
        this.players.delete(oldPlayerId);

        // 更新玩家的 id
        player.id = newPlayerId;

        // 用新的 key 重新添加
        this.players.set(newPlayerId, player);

        // 更新 playerOrder
        const index = this.playerOrder.indexOf(oldPlayerId);
        if (index !== -1) {
            this.playerOrder[index] = newPlayerId;
        }

        Logger.info('PlayerManager', `Player ID updated: ${oldPlayerId} -> ${newPlayerId}`);
        return true;
    }
}

File: poker_arena_server_go/game/player_manager.go
package game

import "poker-arena-server/util"

// PlayerManager manages an ordered collection of players
type PlayerManager struct {
	players     map[string]*TheDecreePlayer
	playerOrder []string
}

func NewPlayerManager() *PlayerManager {
	return &PlayerManager{
		players:     make(map[string]*TheDecreePlayer),
		playerOrder: []string{},
	}
}

func (pm *PlayerManager) CreatePlayers(infos []PlayerInfo) {
	pm.Clear()
	for _, info := range infos {
		p := NewTheDecreePlayer(info)
		pm.players[p.ID] = p
		pm.playerOrder = append(pm.playerOrder, p.ID)
	}
	util.Info("PlayerManager", "Created %d players", len(pm.players))
}

func (pm *PlayerManager) GetPlayer(id string) *TheDecreePlayer {
	return pm.players[id]
}

func (pm *PlayerManager) GetAllPlayers() []*TheDecreePlayer {
	result := make([]*TheDecreePlayer, 0, len(pm.playerOrder))
	for _, id := range pm.playerOrder {
		if p, ok := pm.players[id]; ok {
			result = append(result, p)
		}
	}
	return result
}

func (pm *PlayerManager) GetPlayerOrder() []string {
	out := make([]string, len(pm.playerOrder))
	copy(out, pm.playerOrder)
	return out
}

func (pm *PlayerManager) GetPlayerCount() int {
	return len(pm.players)
}

func (pm *PlayerManager) Clear() {
	pm.players = make(map[string]*TheDecreePlayer)
	pm.playerOrder = []string{}
}

func (pm *PlayerManager) UpdatePlayerID(oldID, newID string) bool {
	p, ok := pm.players[oldID]
	if !ok {
		return false
	}
	delete(pm.players, oldID)
	p.ID = newID
	pm.players[newID] = p

	for i, id := range pm.playerOrder {
		if id == oldID {
			pm.playerOrder[i] = newID
			break
		}
	}
	util.Info("PlayerManager", "Player ID updated: %s -> %s", oldID, newID)
	return true
}
